File: demos/02-agent-service/04-agent-framework/sk-orchestration-py/devops_plugin.py
from datetime import datetime
import textwrap
from typing import Any, Callable, Set

def append_to_log_file(filepath: str, content: str) -> None:
    """Helper function to append content to a log file."""
    with open(filepath, 'a', encoding='utf-8') as file:
        file.write('\n' + textwrap.dedent(content).strip())

def restart_service(service_name: str = "", logfile: str = "") -> str:
    """A function that restarts the named service.
    
    Args:
        service_name: The name of the service to restart
        logfile: The path to the log file
        
    Returns:
        Confirmation message
    """
    log_entries = [
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] ALERT  DevopsAssistant: Multiple failures detected in {service_name}. Restarting service.",
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] INFO  {service_name}: Restart initiated.",
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] INFO  {service_name}: Service restarted successfully.",
    ]

    log_message = "\n".join(log_entries)
    append_to_log_file(logfile, log_message)
    
    return f"Service {service_name} restarted successfully."

def rollback_transaction(logfile: str = "") -> str:
    """A function that rollsback the transaction.
    
    Args:
        logfile: The path to the log file
        
    Returns:
        Confirmation message
    """
    log_entries = [
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] ALERT  DevopsAssistant: Transaction failure detected. Rolling back transaction batch.",
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] INFO   TransactionProcessor: Rolling back transaction batch.",
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] INFO   Transaction rollback completed successfully.",
    ]

    log_message = "\n".join(log_entries)
    append_to_log_file(logfile, log_message)
    
    return "Transaction rolled back successfully."

def redeploy_resource(resource_name: str = "", logfile: str = "") -> str:
    """A function that redeploys the named resource.
    
    Args:
        resource_name: The name of the resource to redeploy
        logfile: The path to the log file
        
    Returns:
        Confirmation message
    """
    log_entries = [
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] ALERT  DevopsAssistant: Resource deployment failure detected in '{resource_name}'. Redeploying resource.",
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] INFO   DeploymentManager: Redeployment request submitted.",
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] INFO   DeploymentManager: Service successfully redeployed, resource '{resource_name}' created successfully.",
    ]

    log_message = "\n".join(log_entries)
    append_to_log_file(logfile, log_message)
    
    return f"Resource '{resource_name}' redeployed successfully."

def increase_quota(logfile: str = "") -> str:
    """A function that increases the quota.
    
    Args:
        logfile: The path to the log file
        
    Returns:
        Confirmation message
    """
    log_entries = [
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] ALERT  DevopsAssistant: High request volume detected. Increasing quota.",
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] INFO   APIManager: Quota increase request submitted.",
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] INFO   APIManager: Quota successfully increased to 150% of previous limit.",
    ]

    log_message = "\n".join(log_entries)
    append_to_log_file(logfile, log_message)

    return "Successfully increased quota."

def escalate_issue(logfile: str = "") -> str:
    """A function that escalates the issue.
    
    Args:
        logfile: The path to the log file
        
    Returns:
        Confirmation message
    """
    log_entries = [
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] ALERT  DevopsAssistant: Cannot resolve issue.",
        f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] ALERT  DevopsAssistant: Requesting escalation.",
    ]
    
    log_message = "\n".join(log_entries)
    append_to_log_file(logfile, log_message)
    
    return "Submitted escalation request."

# Define the set of devops functions
devops_functions: Set[Callable[..., Any]] = {
    restart_service,
    rollback_transaction,
    redeploy_resource,
    increase_quota,
    escalate_issue
}

File: demos/03-agent-framework/04-orchestration-workflow/agentfw-orchestration/Plugins/DevOpsPlugin.cs
using Azure.AI.Agents.Persistent;
using System.Text.Json;

namespace AFWOrchestration;

public static class DevopsPlugin
{
    public static string OutcomeDirectory { get; set; } = string.Empty;

    private static void AppendToLogFile(string filepath, string content)
    {
        // Write to progress log file in outcome directory to track actions as they happen
        var fileName = Path.GetFileName(filepath);
        var progressLogPath = Path.Combine(OutcomeDirectory, fileName.Replace(".log", "-progress.log"));

        // Ensure outcome directory exists
        Directory.CreateDirectory(OutcomeDirectory);

        File.AppendAllText(progressLogPath, "\n" + content.Trim());
    }

    public static string RestartService(string serviceName = "", string logfile = "")
    {
        var logEntries = new[]
        {
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] ALERT  DevopsAssistant: Multiple failures detected in {serviceName}. Restarting service.",
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] INFO  {serviceName}: Restart initiated.",
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] INFO  {serviceName}: Service restarted successfully."
        };

        var logMessage = string.Join("\n", logEntries);
        AppendToLogFile(logfile, logMessage);

        return $"Service {serviceName} restarted successfully.";
    }

    public static string RollbackTransaction(string logfile = "")
    {
        var logEntries = new[]
        {
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] ALERT  DevopsAssistant: Transaction failure detected. Rolling back transaction batch.",
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] INFO   TransactionProcessor: Rolling back transaction batch.",
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] INFO   Transaction rollback completed successfully."
        };

        var logMessage = string.Join("\n", logEntries);
        AppendToLogFile(logfile, logMessage);

        return "Transaction rolled back successfully.";
    }

    public static string RedeployResource(string resourceName = "", string logfile = "")
    {
        var logEntries = new[]
        {
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] ALERT  DevopsAssistant: Resource deployment failure detected in '{resourceName}'. Redeploying resource.",
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] INFO   DeploymentManager: Redeployment request submitted.",
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] INFO   DeploymentManager: Service successfully redeployed, resource '{resourceName}' created successfully."
        };

        var logMessage = string.Join("\n", logEntries);
        AppendToLogFile(logfile, logMessage);

        return $"Resource '{resourceName}' redeployed successfully.";
    }

    public static string IncreaseQuota(string logfile = "")
    {
        var logEntries = new[]
        {
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] ALERT  DevopsAssistant: High request volume detected. Increasing quota.",
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] INFO   APIManager: Quota increase request submitted.",
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] INFO   APIManager: Quota successfully increased to 150% of previous limit."
        };

        var logMessage = string.Join("\n", logEntries);
        AppendToLogFile(logfile, logMessage);

        return "Successfully increased quota.";
    }

    public static string EscalateIssue(string logfile = "")
    {
        var logEntries = new[]
        {
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] ALERT  DevopsAssistant: Cannot resolve issue.",
            $"[{DateTime.Now:yyyy-MM-dd HH:mm:ss}] ALERT  DevopsAssistant: Requesting escalation."
        };

        var logMessage = string.Join("\n", logEntries);
        AppendToLogFile(logfile, logMessage);

        return "Submitted escalation request.";
    }

    public static List<FunctionToolDefinition> GetToolDefinitions()
    {
        return new List<FunctionToolDefinition>
        {
            new FunctionToolDefinition(
                name: nameof(RestartService),
                description: "A function that restarts the named service",
                parameters: BinaryData.FromObjectAsJson(new
                {
                    Type = "object",
                    Properties = new
                    {
                        ServiceName = new { Type = "string", Description = "The name of the service to restart" },
                        Logfile = new { Type = "string", Description = "The path to the log file" }
                    },
                    Required = new[] { "serviceName", "logfile" }
                },
                new JsonSerializerOptions() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase })
            ),
            new FunctionToolDefinition(
                name: nameof(RollbackTransaction),
                description: "A function that rolls back the transaction",
                parameters: BinaryData.FromObjectAsJson(new
                {
                    Type = "object",
                    Properties = new
                    {
                        Logfile = new { Type = "string", Description = "The path to the log file" }
                    },
                    Required = new[] { "logfile" }
                },
                new JsonSerializerOptions() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase })
            ),
            new FunctionToolDefinition(
                name: nameof(RedeployResource),
                description: "A function that redeploys the named resource",
                parameters: BinaryData.FromObjectAsJson(new
                {
                    Type = "object",
                    Properties = new
                    {
                        ResourceName = new { Type = "string", Description = "The name of the resource to redeploy" },
                        Logfile = new { Type = "string", Description = "The path to the log file" }
                    },
                    Required = new[] { "resourceName", "logfile" }
                },
                new JsonSerializerOptions() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase })
            ),
            new FunctionToolDefinition(
                name: nameof(IncreaseQuota),
                description: "A function that increases the quota",
                parameters: BinaryData.FromObjectAsJson(new
                {
                    Type = "object",
                    Properties = new
                    {
                        Logfile = new { Type = "string", Description = "The path to the log file" }
                    },
                    Required = new[] { "logfile" }
                },
                new JsonSerializerOptions() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase })
            ),
            new FunctionToolDefinition(
                name: nameof(EscalateIssue),
                description: "A function that escalates the issue",
                parameters: BinaryData.FromObjectAsJson(new
                {
                    Type = "object",
                    Properties = new
                    {
                        Logfile = new { Type = "string", Description = "The path to the log file" }
                    },
                    Required = new[] { "logfile" }
                },
                new JsonSerializerOptions() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase })
            )
        };
    }
}
